service/history/historyEngine.go

```diff
@@ -452,16 +452,15 @@ Update_History_Loop:
 			return nil, err
 		}
 		timerTasks = append(timerTasks, start2CloseTimeoutTask)
-		defer e.timerProcessor.NotifyNewTimer(start2CloseTimeoutTask.GetTaskID())
 
 		start2HeartBeatTimeoutTask, err := context.tBuilder.AddHeartBeatActivityTimeout(ai)
 		if err != nil {
 			return nil, err
 		}
 		if start2HeartBeatTimeoutTask != nil {
 			timerTasks = append(timerTasks, start2HeartBeatTimeoutTask)
-			defer e.timerProcessor.NotifyNewTimer(start2HeartBeatTimeoutTask.GetTaskID())
 		}
+		defer e.timerProcessor.NotifyNewTimer(timerTasks)
```

Comment:
Same as above.  All notifications should make sure transactions goes through before notifying.